framework/include/utils/IndexableProperty.h

```diff
@@ -0,0 +1,125 @@
+                    const std::string & component_param = "component");
+
+  GenericReal<is_ad> operator[](int i) const;
+  operator bool() const;
+
+protected:
+  template <typename P>
+  void getPropertyHelper(const GenericMaterialProperty<P, is_ad> *& pointer,
+                         unsigned int components);
+
+  const GenericMaterialProperty<Real, is_ad> * _mat_prop_real;
+  const GenericMaterialProperty<RealVectorValue, is_ad> * _mat_prop_real_vector_value;
+  const GenericMaterialProperty<RankTwoTensor, is_ad> * _mat_prop_rank_two_tensor;
+  const GenericMaterialProperty<RankThreeTensor, is_ad> * _mat_prop_rank_three_tensor;
+  const GenericMaterialProperty<RankFourTensor, is_ad> * _mat_prop_rank_four_tensor;
+
+  T * _host;
+
+  const std::string & _mat_prop_param;
+  const std::string & _mat_prop_name;
+  const std::string & _component_param;
+
+  const std::vector<unsigned int> _component;
+};
+
+template <typename T, bool is_ad>
+InputParameters
+IndexableProperty<T, is_ad>::validParams()
+{
+  auto params = T::validParams();
+  params.template addRequiredParam<MaterialPropertyName>("mat_prop",
+                                                         "The name of the material property");
+  params.template addParam<std::vector<unsigned int>>(
+      "component",
+      "Index vector of the scalar component to extract from "
+      "the material property (empty for scalar properties)");
+  return params;
+}
+
+template <typename T, bool is_ad>
+IndexableProperty<T, is_ad>::IndexableProperty(T * host,
+                                               const std::string & mat_prop_param,
+                                               const std::string & component_param)
+  : _host(host),
+    _mat_prop_param(mat_prop_param),
+    _mat_prop_name(_host->template getParam<MaterialPropertyName>(_mat_prop_param)),
+    _component_param(component_param),
+    _component(host->template getParam<std::vector<unsigned int>>(_component_param))
+{
+  getPropertyHelper<Real>(_mat_prop_real, 0);
+  getPropertyHelper<RealVectorValue>(_mat_prop_real_vector_value, 1);
+  getPropertyHelper<RankTwoTensor>(_mat_prop_rank_two_tensor, 2);
+  getPropertyHelper<RankThreeTensor>(_mat_prop_rank_three_tensor, 3);
+  getPropertyHelper<RankFourTensor>(_mat_prop_rank_four_tensor, 4);
+}
+
+template <typename T, bool is_ad>
+GenericReal<is_ad>
+IndexableProperty<T, is_ad>::operator[](int i) const
+{
+  if (_mat_prop_real)
+    return (*_mat_prop_real)[i];
+  if (_mat_prop_real_vector_value)
+    return (*_mat_prop_real_vector_value)[i](_component[0]);
+  if (_mat_prop_rank_two_tensor)
+    return (*_mat_prop_rank_two_tensor)[i](_component[0], _component[1]);
+  if (_mat_prop_rank_three_tensor)
+    return (*_mat_prop_rank_three_tensor)[i](_component[0], _component[1], _component[2]);
+  if (_mat_prop_rank_four_tensor)
+    return (*_mat_prop_rank_four_tensor)[i](
+        _component[0], _component[1], _component[2], _component[3]);
+  mooseError(
```

Comment:
It seems like this error should also be in the constructor so it's hit right away. If you did that, it wouldn't really be necessary here.

Reply:
That would add more code for the check (both written _and_ executed).

Reply:
still better to error at construction time when possible imo